### ISSUE TITLE
Websocket `idleTimeout` leads to Broken pipe

### ISSUE THREAD

**Issue by irbull** [denoland/deno#15091]
Tested on Deno 🦖 1.23.3

If I create a WS Server and set an `idleTimeout`, if that timeout is reached the connection is closed but the client doesn't seem to know this. If the client tries to `send` data, the Deno runtime crashes with the following error:

```
Error 'IO error: Broken pipe (os error 32)' contains boxed error of unknown type:
  Io(Os { code: 32, kind: BrokenPipe, message: "Broken pipe" })
  Os { code: 32, kind: BrokenPipe, message: "Broken pipe" }
```

Here is a snippet that can reproduce this:

```
// ws_server.ts
import { deferred } from "https://deno.land/std@0.144.0/async/deferred.ts";

const errorDeferred = deferred();
const closeDeferred = deferred();

const listener = Deno.listen({ port: 4509 });
console.log("READY");
const httpConn = Deno.serveHttp(await listener.accept());
const { request, respondWith } = (await httpConn.nextRequest())!;
const { response, socket } = Deno.upgradeWebSocket(request, {
  idleTimeout: 1,
});

socket.onmessage = (message: MessageEvent) => {
  console.log(`message: ${message.data}`);
};

socket.onerror = (e) => {
  console.log(`error received: ${(e as ErrorEvent).message}`);
  errorDeferred.resolve();
};
socket.onclose = (e) => {
  console.log(`closed received, ${e.reason}`);
  closeDeferred.resolve();
};
await respondWith(response);

await errorDeferred;
await closeDeferred;
listener.close();
```

and the client:
```
const ws = new WebSocket("ws://localhost:4509");
ws.onopen = () => {
  // Do something that takes a while
  for (let i = 0; i < 10000; i++) {
    Deno.writeFileSync("testFile", new TextEncoder().encode("hello world"));
  }
  try {
    console.log(
      `send request now, but it's likely closed. WS state: ${getReadyState(ws.readyState)}`,
    );
    ws.send("testdata");
  } catch (e) {
    console.log(e);
  }
};

function getReadyState(readyState: number): string {
  switch (readyState) {
    case WebSocket.CONNECTING:
      return "connecting";
    case WebSocket.OPEN:
      return "open";
    case WebSocket.CLOSING:
      return "closing";
    case WebSocket.CLOSED:
      return "closed";
    default:
      throw new Error(`Invalid websocket readyState: ${readyState}`);
  }
}
```

Start the server with: `deno run -A ws_server.ts` and start the client with `deno run -A ws_client.ts`.

The client will create a WS Connection, and then do something that takes longer than 1 second. It then tries to send something on the websocket. This fails, which would be ok if the exception could be caught. However, the entire runtime just crashes.

**Comment by irbull** [comment]
If the client attempts to get the `readyState` it says it's `open`, so we can't even preemptively test the state of the websocket reliably.

**Comment by littledivy** [comment]
This was fixed in the latest release (1.36.1). Race condition in the recv buffer.

**Comment by soootaleb** [comment]
Wonderful @littledivy thanks a lot, you guys are building something awesome.

Can I ask you to share the commit that fixed the root cause please, I'm having a hard time finding a reference in the issues mentioned around here.

Regards
